### DISCUSSION TITLE
Is there any way to change drivers timezone to proxy location timezone in UC Mode?

### DISCUSSION THREAD

**Discussion by suredarv06** [seleniumbase/SeleniumBase#2464] [General]
In UC mode is there any option for changing timezone of browser to the proxy location timezone.

**Comment by mdmintz** [comment]
Duplicate of https://github.com/seleniumbase/SeleniumBase/issues/918#issuecomment-864054769

**Reply by suredarv06** [reply]
```from seleniumbase import Driver


driver = Driver(uc=True,uc_cdp=True)
try:
    tz_params = {'timezoneId': 'America/Costa_Rica' }
    driver.execute_cdp_cmd('Emulation.setTimezoneOverride', tz_params)
    driver.get("https://pixelscan.net/")
    driver.sleep(100)
finally:
    driver.quit()```


I have Overridden timezone ,
But (https://pixelscan.net/),Finds my correct timezone

**Reply by mdmintz** [reply]
[Emulation.setTimezoneOverride](https://chromedevtools.github.io/devtools-protocol/tot/Emulation/#method-setTimezoneOverride) comes directly from the Chrome DevTools Protocol. If you have an issue with that method, reach out to Google directly. 🙂

**Reply by slvamegestic** [reply]
The same thing is happening to me as well: https://pixelscan.net finds me well, I think Google has updated their security. There is no prettier way to override the timezone than  [Emulation.setTimezoneOverride](https://chromedevtools.github.io/devtools-protocol/tot/Emulation/#method-setTimezoneOverride). Is there anything the Selenium base driver can do instead?

The question isn't about setting the timezone, but rather overriding it with this superior undetectable selenium base driver without detecting by a bot.

**Reply by mdmintz** [reply]
Overriding the timezone isn't necessary to avoid being detected:

```python
from seleniumbase import SB

with SB(uc=True) as sb:
    sb.driver.uc_open_with_reconnect("https://pixelscan.net", 20)
```

<img width="1258" alt="Screenshot 2024-02-06 at 9 11 51 AM" src="https://github.com/seleniumbase/SeleniumBase/assets/6788579/9cb16a91-b068-4a0e-a8c9-a243878c8556">

**Reply by slvamegestic** [reply]
That's cool,
But in my case, I need to update the timezone my proxy_server.
```
from seleniumbase import SB
sb_config={}
sb_config['proxy']='<my-proxy>'  
sb_config['uc']=True
with SB(**sb_config) as sb:
 
    tz_params = { 'timezoneId': 'America/Costa_Rica' }
    sb.driver.execute_cdp_cmd('Emulation.setTimezoneOverride', tz_params)
    sb.driver.uc_open_with_reconnect("https://pixelscan.net", 2000)
```
The above shows that I am trying to set my proxy timezone, but it still says that I am spoofing the location.

**The problem is that I need to bypass the location even through the proxy server. How can I do this?**

<img width="1166" alt="Screenshot 2024-02-06 at 8 09 32 PM" src="https://github.com/seleniumbase/SeleniumBase/assets/35224544/fc1a7092-bf8c-4813-9815-0444e6d55a19">

**Reply by mdmintz** [reply]
@slvamegestic That's up to the proxy server. They can get your actual location via the IP Address.

**Reply by slvamegestic** [reply]
That's cool @mdmintz, I'm wondering if there's a feature in any way to overwrite that in the Selenium base driver or other on how we can ensure that the Selenium base driver doesn't leak IP information? This is crucial for maintaining security and privacy, especially when dealing with sensitive tasks. 
Already go through the discussion [How can i hide my real ip](https://github.com/seleniumbase/SeleniumBase/discussions/2276)
Here my IP looks good and reputated but the location is leaking.

Any help or suggestions in this regard would be greatly appreciated. Thanks

**Reply by suredarv06** [reply]
`[Emulation.setTimezoneOverride](https://chromedevtools.github.io/devtools-protocol/tot/Emulation/#method-setTimezoneOverride) `
It is google's api

Ok fine, In my case I am using proxy. So I need to change my timezone to proxy timezone. Emulation.setTimezoneOverride is changing my browser timezone which is reflected in browserleaks , But pixelscan finds my exact timezone after overriding also.

1. I don't know how it finds ,Do you have idea?  If any share with me
2. Any other option available in seleniumBase to overcome that?

**Reply by mdmintz** [reply]
Your real IP Address is not leaked if you use a proxy: https://github.com/seleniumbase/SeleniumBase/issues/2217#issuecomment-1784010586
(Then, only your proxy's IP Address is shown.)

**Reply by slvamegestic** [reply]
@mdmintz I've already dealt with that issue. In this case, the issue is only with [pixelscan](https://pixelscan.net/).

[https://browserleaks.com/](url) and even [http://f.vision/](url) passed my test.

<img width="1166" alt="Screenshot 2024-02-07 at 2 41 04 PM" src="https://github.com/seleniumbase/SeleniumBase/assets/35224544/02006e1a-e843-472b-a04d-8d326bd2f370">

There appears to be a leak in the IP with only pixelscan. So perhaps it has something to do with undetectable chrome and IP address.
Any help or suggestions in this regard would be greatly appreciated. Thanks

**Reply by mdmintz** [reply]
Don't know about that, but it does say "No automation framework detected", which is really the only part that matters.

<img width="1267" alt="Screenshot 2024-02-07 at 9 06 21 AM" src="https://github.com/seleniumbase/SeleniumBase/assets/6788579/2bd72035-85a9-4737-903e-97d7282f2a73">

**Reply by guryaniv** [reply]
Hi, I'm experiencing the same issue.
Undetectable mode does work without a proxy, but when setting a proxy I'm getting detected.
When running pixelscan I get "Look like you are spoofing your location" because my IP (proxy) timezone is different than my Javascript (actual) timezone.
Also the proxy is rotating, so it will require to check the ip location, find the matching timezoneId and use Emulation.setTimezoneOverride on each request.
[@mdmintz](https://github.com/mdmintz) Maybe do you know of any easier way to implement this?
I'd appreciate your help.

**Reply by slvamegestic** [reply]
@guryaniv want to know, 
Does setting a proxy IP timezone ID using Emulation.setTimezoneOverride prevent location spoofing?

**Reply by guryaniv** [reply]
@slvamegestic looks like it works, at least initially. when I init the driver and then use Emulation.setTimezoneOverride with the proxy IP timezone I pass the pixelscan check
![image](https://github.com/seleniumbase/SeleniumBase/assets/24723031/43f8cccf-41d8-4720-b942-3df11e49c88d)

**Reply by slvamegestic** [reply]
@guryaniv  Could you please share the sample code that initially worked for you?

**Reply by Vrackko** [reply]
My code with Emulation.setTimezoneOverride sometimes work sometimes dont on both pixelscan and fake vision. Here is sample of my code maybe you guys can figure it out (Also chaning header doesnt work sometimes):

`
driver = Driver(uc=True, agent=user_agent, extension_dir=extension_path2, incognito=True, proxy=proxy_cred )
try:
    
    header_params = {
        'headers': {
            'sec-ch-ua-platform': 'Android',
            'sec-ch-ua-mobile': '?1'
        }
    }
    driver.execute_cdp_cmd('Network.setExtraHTTPHeaders', header_params)
    
    tz_params = {'timezoneId': 'America/New_York' }
    driver.execute_cdp_cmd('Emulation.setTimezoneOverride', tz_params)

    driver.get("http://f.vision/")   #   https://browserleaks.com/ip  https://pixelscan.net    http://f.vision/
 
    driver.sleep(1000)
finally:
    driver.quit()
`

**Reply by Vrackko** [reply]
I think i found the problem. Thing is pixelscan scans your data/time fast so your code doesnt have time to inject new date/time timezone. After refreshing you should get right date/time timzone.

**Reply by Ishmohit** [reply]
Hi @slvamegestic , I'm looking into the same issue of setting a timezone override while connecting through proxy. Were you able to find a solution for this?

**Reply by slvamegestic** [reply]
@Ishmohit  Still, pixelscan and other antibot websites able to detect my timezone spoofing, but other metrics passed by Seleniumbase, so the profile looks good. I am tired of find a solution for the timezone issue it still persists for me
